### PR TITLE
NumberFields should allow numbers with decimals to be entered

### LINKED ISSUE #2214: NumberFields should allow numbers with decimals to be entered
Jenn found a bug in #2174 where the FTE field won't let you put in a decimal point.  What happens is that the onChange event formats the entry into a number, which strips off leading zeros and trailing decimal points.  (fun fact, if you really want a number like 0.6 you can start with 106, make it into 10.6, and remove the leading 1 ... not the best user experience).  So when I investigated I realized that the existing FTE field in the State Cost Categories doesn't have that problem because it formats FTE's as a string.  But that means that you can put in bad characters like letters and it won't complain.  

DollarField has a pretty neat solution for this where it separates out a changeHandler to allow each character as you type it and blurHandler that does the final conversion into a string.  To fix this issue, move (or copy) that functionality to NumberField (move as much as possible so as not to have duplicate code, but keep separate whatever DollarField needs to handle it's own unique formatting requirements)

### This task is done when...
 - [ ] NumberField allows numbers with decimals to be entered

### PR DESCRIPTION
**Pull request by pphillips-fearless**
Fix bug where FTE fields don't allow you to enter a decimal point.  This task refactored NumberField to handle number fields by moving and enhancing DollarField change handling code

### This pull request changes...

- number fields can handle decimal values
- decimal fields still work as they used to
- percent fields are still rounded to the nearest integer
- number fields gracefully handle non-numeric characters
- closes #2214 

### This pull request is ready to merge when...

- [ ] Tests have been updated (and all tests are passing)
- [ ] This code has been reviewed by someone other than the original author
- [ ] The experience passes a basic manual accessibility audit (keyboard nav, screenreader, text scaling) OR an exemption is documented
~~- [ ] The change has been documented~~
~~- [ ] Associated OpenAPI documentation has been updated~~
- [ ] Changelog is updated as appropriate

### CODE REVIEW

**Comment by jeromeleecms** [comment]
Can you post a screenshot where I should be validating this? #2174 is related to the Key Personnel, but that is still in % when I go to the deploy site... or is this a broader FTE field under activities...

**Comment by mgwalker** [comment]
@jeromeleecms This should probably be held until #2202 is merged, since that's where the underlying issue originated. We can't actually verify that it's fixed if the original issue hasn't presented yet. 😄

**Comment by pphillips-fearless** [comment]
Just to answer @jeromeleecms and give a couple of screenshots...

There are lots of places where number entry fields are affected by this change:
1) Dollar fields are many places and they all need to still work as before (with commas for large number and rounding to nearest dollar, and they shouldn't crash if invalid characters are entered)
2) Generic number fields are on incentive payments as EH counts and EP counts and on contractor resource form as number of hours when a private contractor cost is hourly in the activities.  They should handle commas and decimal points and shouldn't crash if invalid characters are entered.
3) FTEs are on state personnel form in the State Cost Categories page in the activities.  They should handle commas and decimal points and shouldn't crash if invalid characters are entered.
4) Percents fields are on the key personnel (before it gets changed in 2174), quarterly expenditure on the FFP and budget page.  They should round to the nearest percent and shouldn't crash if invalid characters are entered.
5) Note: dates are not affected by this change.

Screen Shot 2020-05-15 at 10 24 38 AM (2)

Screen Shot 2020-05-15 at 10 35 26 AM (2)

**Comment by pphillips-fearless** [comment]
@jeromeleecms I figured out why we saw the error on the key personnel percent commitment to program number when we tested together earlier.  I originally thought it was because of the comma, but it was actually because the apd schema had a regex pattern on it that required the number to be <= 100%.  Now that we're using FTE the regex is gone so it won't be an issue.  I searched the schema for other similar regex patterns and didn't find any on any data entry numbers.

**Comment by pphillips-fearless** [comment]
@mgwalker, @jeromeleecms the changes from 2174 are merged in here and it's ready to review again.

**Comment by jeromeleecms** [comment]
@pphillips-fearless -- If the field rounds, should the math also be rounding? 

I put in some really odd numbers and the math rounded, but the calculation came through:

image

When collapsed, the number which rounded to zero is showing up in the FTE field:

image

**Comment by pphillips-fearless** [comment]
@jeromeleecms , I figured out the cause of the rounding issue you reported.  The javascript function that puts the commas in the dollar fields (toLocaleString) also specifies a maximum number of fraction digits of 3 before rounding. So the stored number might be 0.00001 but the displayed number would be zero.  The default of 3 digits is configurable, but whatever you set it to will cause problems in extreme situations.  So I recommend we pick a max number of fraction digits (3 would be easiest because it's the default for toLocaleString) and round any numeric value we actually store once we get to that number of fraction digits.

**Comment by pphillips-fearless** [comment]
Per parking lot discussion today, we will set it to round past 4 decimal places.  So, for example, we will store 0.0001, but if we get 0.00001 it will round to zero.  In all cases, what gets stored and what gets displayed will be the same number.

**Comment by jeromeleecms** [comment]
To also document the rational for the rounding:

1) An FTE reports at ~1800 - 2080 hours per year. 
2) If we used 0.0001 to draw the line, that puts us at reporting fidelity between 10-12 minute blocks (1800 hours * 0.0001 * 60 minutes), which per our discussion in parking lot time is probably more granular than staff might report work (which is done in 1/4 hour or 15 minute intervals).
3) This should address the lower end extreme of a single individual as well as people greater than 1.
4) We will address limits on fields in #2054 and any resulting issues.

**Comment by pphillips-fearless** [comment]
Just to document my comment at our meeting this afternoon, when I implemented the change from Greg's code review comment above, it ended up removing the toLocaleString function, which took away the rounding issue.  But then as I tested I saw that there are still math issues when multiplying very large numbers by numbers with large number of decimal places, so I still think we should go with our decision to limit numbers to 4 decimal places.

**Comment by pphillips-fearless** [comment]
@mgwalker I think I got all your comments, ready for you to take another look

File: web/src/components/NumberField.test.js [hunk 1]
```diff
@@ -20,6 +20,51 @@ describe('NumberField component', () => {
...
+          label="test label"
+          name="test name"
+          size="medium"
+          className="stuff"
+          value="abc"
+          onChange={jest.fn()}
+        />
+      )
+    ).toMatchSnapshot();
+  });
```
[the remark this answers is not in the document]
Comment (pphillips-fearless) [reply web/src/components/NumberField.test.js:36]: This was moved over from the old DollarField tests, I thought the case of non-numeric initial values would still be applicable.  I guess the last one, where we show that non-numeric values are handled, is a better way to show it.

File: web/src/components/NumberField.test.js [hunk 2]
```diff
@@ -20,6 +20,51 @@ describe('NumberField component', () => {
...
+          name="test name"
+          size="medium"
+          className="stuff"
+          value="123321"
+          onChange={jest.fn()}
+        />
+      )
+    ).toMatchSnapshot();
+  });
+
```
[the remark this answers is not in the document]
Comment (pphillips-fearless) [reply web/src/components/NumberField.test.js:67]: makes sense

**Pull request merged**